Advantage computation in the RefActor

Darktex opened the pull request:
Run with:

```
python scripts/runnable_recipe_ray_vllm_weight_sync.py --config ../recipes/configs/dev/qwen3B_full_grpo.yaml
```

Things look good using the "default" arguments: 

```
# Fine-tuning arguments
batch_size: 16 # B*G
grpo_samples: 16
forward_batch_size: 1
max_generated_tokens: 512

# number of train_steps after which to do weight sync
steps_before_sync: 2
num_ref_workers: 1
num_fsdp_workers: 2

vllm:
  num_workers: 5
  tp_size: 1
  batch_size: 1 # TODO: for now, needs to match recipe batch size
  queue_maxsize: 3 #num workers * steps_before_sync
```
<img width="1830" alt="Screenshot 2025-03-31 at 5 59 07 PM" src="https://github.com/user-attachments/assets/3214ed34-d785-4a69-bf45-6ec036e60f00" />
<img width="1819" alt="Screenshot 2025-03-31 at 5 59 33 PM" src="https://github.com/user-attachments/assets/3d171f02-c129-4845-ba76-304e6626ccba" />

felipemello1 commented on recipes/configs/dev/qwen3B_full_grpo.yaml:
```diff
@@ -71,8 +71,8 @@ resume_from_checkpoint: False
 # Fine-tuning arguments
-batch_size: 1
-grpo_samples: 16
+batch_size: 6
+grpo_samples: 5
```
1. Shouldnt we have one for RefActor too?
2. super cool with leaving it like this, but at some point we might want to do what we did for vllm, and have it actor have its own section

Darktex replied:
Great call. Makes sense to me, perhaps this can even be applied at the actor level (once we start factoring out common components?)

felipemello1 commented on scripts/runnable_recipe_ray_vllm_weight_sync.py:
```diff
@@ -274,39 +284,24 @@ def run(self):
                     if self._is_actor_zero:
                         print(f"Getting from rollout_queue queue.")
                     trajectory = self.rollout_queue.get(timeout=0.5)
+                    trajectory = trajectory.to(self._device)
```
most likely, this is optimal, but we may not need for every content to be in the gpu, e.g metadata.

Darktex replied:
GPUs can't hold strings anyway, so that should be fine

felipemello1 commented on scripts/runnable_recipe_ray_vllm_weight_sync.py:
```diff
@@ -274,39 +284,24 @@ def run(self):
+                trajectory["query_response_padding_masks"]
             )
 
             # Reset GPU memory stats before model_running
```
i think we can delete this. Our utility already resets it.

felipemello1 commented on scripts/runnable_recipe_ray_vllm_weight_sync.py:
```diff
@@ -315,33 +310,89 @@ def run(self):
             with torch.no_grad():
                 ref_logits = self._ref_model(
-                    query_responses, input_pos=position_ids, mask=masks
+                    trajectory["query_responses"], input_pos=position_ids, mask=masks
```
did you decide to keep it named as "query_responses"? should we change to something more obvious?

Darktex replied:
Let's do this as part of a refactor

felipemello1 commented on scripts/runnable_recipe_ray_vllm_weight_sync.py:
```diff
@@ -315,33 +310,89 @@ def run(self):
             del ref_logits, position_ids, masks
             # masking of ref_logprobs is done in grpo_step
 
+            # Extract components from raw trajectory
+            query_responses = trajectory["query_responses"]
+            responses = trajectory["responses"]
+            query_response_padding_masks = trajectory["query_response_padding_masks"]
+            seq_lens = trajectory["seq_lens"]
+            answers = trajectory["answers"]
```
if we are mostly doing trajectory["key"] in the code before it, should we keep the same pattern and avoid unwrapping the trajectory?

Darktex replied:
Yeah but we need to refactor this whole portion anyway :D

felipemello1 commented on scripts/runnable_recipe_ray_vllm_weight_sync.py:
```diff
@@ -315,33 +310,89 @@ def run(self):
+            response_padding_masks, responses = (
+                rlhf.truncate_sequence_at_first_stop_token(
+                    responses,
+                    torch.tensor(self._tokenizer.stop_tokens, device=self._device),
```
creating this tensor every time seems wasteful. Maybe we can move it elsewhere?

felipemello1 commented on scripts/runnable_recipe_ray_vllm_weight_sync.py:
```diff
@@ -315,33 +310,89 @@ def run(self):
+                self._tokenizer, responses, answers, device=self._device
+            )
+
+            # Compute advantages: B, G, num_funcs -> B, G
```
there was an all_reduce function here before, for the sake of logging the sum of rewards across all actors. Should we migrate the logging of rewards/successes to RefActor?

Darktex replied:
Logging migrated

felipemello1 commented on scripts/runnable_recipe_ray_vllm_weight_sync.py:
```diff
@@ -608,14 +659,14 @@ def postprocess_vllm_request_output(request_output):
-                query_response_padding_masks,
-                seq_lens,
-            ]
+            return TensorDict(
+                query_responses=query_responses,
+                responses=responses,
+                logprobs=logprobs,
+                query_response_padding_masks=query_response_padding_masks,
+                seq_lens=seq_lens,
+                batch_size=bs,
+            )
```
awesome

Darktex replied:
A lot more to do...

felipemello1 commented on scripts/runnable_recipe_ray_vllm_weight_sync.py:
```diff
@@ -1324,11 +1359,15 @@ def train(self):
             )
             time_waiting_buffer = time.perf_counter() - time_waiting_buffer_start
-            log.info(f"{self.rank=} got from queue traj {trajectory}")
+            if self._is_rank_zero:
+                log.info(f"{self.rank=} got from queue traj {trajectory}")
```
we have utils.log_rank_zero, so you dont have to do `if self._is_rank_zero:`

felipemello1 commented on scripts/runnable_recipe_ray_vllm_weight_sync.py:
```diff
@@ -1529,24 +1559,10 @@ def _prepare_trajectory(self, raw_trajectory):
             seq_lens=training.get_unmasked_sequence_lengths(response_padding_masks),
         )
 
-        # Reduce rewards and successes for logging
-        torch.distributed.reduce(rewards_full, dst=0, op=torch.distributed.ReduceOp.AVG)
-        torch.distributed.reduce(
-            successes_full, dst=0, op=torch.distributed.ReduceOp.AVG
-        )
-        if self._is_rank_zero:
-            rewards_mean_per_func = rewards_full.mean(dim=(0, 1)).cpu()
-            successes_mean_per_func = successes_full.mean(dim=(0, 1)).cpu()
-            rewards_mean = rewards_mean_per_func.mean()
-            successes_mean = successes_mean_per_func.mean()
-        else:
-            rewards_mean_per_func = successes_mean_per_func = rewards_mean = (
-                successes_mean
-            ) = None
-
-        # Clean up intermediate variables
-        del rewards_full, successes_full, responses, rewards_sum
-        torch.distributed.barrier()
+        rewards_mean_per_func = rewards.mean(dim=0).cpu()
+        successes_mean_per_func = successes.mean(dim=0).cpu()
+        rewards_mean = rewards_mean_per_func.mean()
+        successes_mean = successes_mean_per_func.mean()
```
this means that the reward we log will be only from rank 0. I am ok with that. But does it make sense to you?

SalmanMohammadi commented on scripts/runnable_recipe_ray_vllm_weight_sync.py:
```diff
@@ -315,33 +310,89 @@ def run(self):
                 )
             time_model_running = time.perf_counter() - time_grpo_steps_start
 
             ref_logits = rlhf.truncate_sequence_for_logprobs(ref_logits, context_length)
```
This assumes batched prompts have been padded to a fixed length so `context_length` is fixed for every sample
```python
def truncate_sequence_for_logprobs(
    query_response_logits: torch.Tensor, context_length: int
) -> torch.Tensor:
    """
    Truncates logits generated over a sequence for estimating logprobs over the tokens in the sequence.
    This assumes the sequence is of the (query, response) format with length (context_length + response_length)
    Args:
        query_response_logits (torch.Tensor): The logits tensor of shape [b, context_length + response_length, vocab_size].
        context_length (int): The length of the context.

    Returns:
        torch.Tensor: The truncated logits for the response with shape [b, response_length, vocab_size]."""
    return query_response_logits[:, context_length - 1 : -1]
```

Darktex replied:
It's already the case, right? We'll do packing later

SalmanMohammadi commented on scripts/runnable_recipe_ray_vllm_weight_sync.py:
```diff
@@ -315,33 +310,89 @@ def run(self):
+
+            # Compute advantages: B, G, num_funcs -> B, G
+            rewards_sum = rewards_full.sum(-1)
+            advantages = (rewards_sum - rewards_sum.mean(1, keepdim=True)) / (
```
Have the rewards already been gathered by this point?

Darktex replied:
There is no torch.distributed group in the RefActor (well, at least not for now), so no need to do that

Darktex commented:
Since last time, made a few changes:

1. Moved the advantage and reward logging into the RefActor accordingly
2. Added more explicit shapes
3. Changed the reward computation function to take answers of size (B, G) and manually duplicated answers. This is a first step to being super deliberate about shapes, with more refactoring to come.

Issues: 

I don't know if it's the logging or what, but I'm not getting good curves from my runs. 

I could use help in checking what's going on here. I followed the same settings as Felipe:

```
# Fine-tuning arguments
batch_size: 1
grpo_samples: 16
forward_batch_size: 1
max_generated_tokens: 512

# number of train_steps after which to do weight sync
steps_before_sync: 2
num_ref_workers: 1
num_fsdp_workers: 2

vllm:
  num_workers: 5
  tp_size: 1
  batch_size: 1 # TODO: for now, needs to match recipe batch size
  queue_maxsize: 3 #num workers * steps_before_sync
```

<img width="1842" alt="Screenshot 2025-03-31 at 5 34 23 PM" src="https://github.com/user-attachments/assets/343a06cc-8df8-4b90-8b3e-b99310c67f32" />

Darktex commented on recipes/configs/dev/qwen3B_full_grpo.yaml:
```diff
@@ -116,7 +115,7 @@ num_fsdp_workers: 2
   num_workers: 5
   tp_size: 1
-  batch_size: 1 # TODO: for now, needs to match recipe batch size
+  batch_size: 1
```
Keeping these values like this for now, but we need to move away from using 1 as a number... like ANYWHERE :D

vmoens reviewed (approved):
LGTM!

SalmanMohammadi commented on scripts/runnable_recipe_ray_vllm_weight_sync.py:
```diff
@@ -274,39 +320,24 @@ def run(self):
-            ) = trajectory
-
-            context_length = query_responses.shape[1] - responses.shape[1]
+            context_length = (
```
When will this not just be `context_length = responses.shape[1]`?
